pkg/kv/kvserver/rafttermcache/term_cache.go

```diff
@@ -0,0 +1,384 @@
+	// TODO(hakuuww): using a RWMutex is one way, and can guarantee safeness.
+	//  But maybe we can just lock replica.mu externally when calling ScanAppend()
+	//  and avoid using locks when calling LoadTerm().
+	mu syncutil.RWMutex
+}
+
+// errInvalidEntryID is returned when the supplied entryID
+// is invalid for the operation. Indicates inconsistent log entries.
+var errInvalidEntryID = errors.New("invalid entry ID")
+
+// NewTermCache initializes a TermCache with a fixed maxSize.
+func NewTermCache(size uint64) *TermCache {
+	return &TermCache{
+		cache:     make([]entryID, 0, size),
+		maxSize:   size,
+		lastIndex: 0,
+	}
+}
+
+// Term returns the entry term based on the given entry index.
+// Returns error if not in the term cache.
+func (tc *TermCache) Term(
+	ctx context.Context, index kvpb.RaftIndex,
+) (term kvpb.RaftTerm, found bool) {
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+
+	if len(tc.cache) == 0 {
+		log.Fatalf(ctx, "term cache is empty when calling Term(), should not happen")
+		return 0, false
+	}
+
+	if index < tc.firstEntry().index ||
+		index > tc.lastIndex {
+		return 0, false
+	}
+
+	// in last term of term cache, index <= tc.lastIndex
+	if index >= tc.lastEntry().index {
+		return tc.lastEntry().term, true
+	}
+
+	for i := len(tc.cache) - 2; i >= 0; i-- {
+		if index >= tc.cache[i].index {
+			return tc.cache[i].term, true
+		}
+	}
+
+	return 0, false
+}
+
+// ClearTo clears entries from the TermCache with index strictly less than hi.
+// If hi is above the lastIndex, the whole term cache is cleared.
+// Mirrors the clearTo function in raftentry/cache.go
+func (tc *TermCache) ClearTo(hi kvpb.RaftIndex) {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+
+	if len(tc.cache) == 0 || hi <= tc.firstEntry().index {
+		return
+	}
+
+	// special cases:
+	// keep the last entry in storage
+	if hi >= tc.lastIndex {
+		tc.resetWithFirstNoLock(tc.lastEntry().term, tc.lastIndex)
+		return
+	}
+
+	// hi is above last entry's index, but lower or equal to lastIndex
+	if hi > tc.lastEntry().index {
+		tc.resetWithFirstNoLock(tc.lastEntry().term, hi)
+		return
+	}
+
+	// only keep the last entry
+	if hi == tc.lastEntry().index {
+		tc.cache = tc.cache[len(tc.cache)-1:]
+		return
+	}
+
+	// general cases
+	for i := 0; i < len(tc.cache)-1; i++ {
+		// hi matches a term flip index
+		if hi == tc.cache[i].index {
+			tc.cache = tc.cache[i:]
+			return
+		}
+
+		// Allow the first entry in the term cache to not represent a term flip point
+		// cache[0] only tells us entries are in term cache[0].term
+		// starting from cache[0].index up to
+		// min(cache[i+1].index-1 /* if not nil*/, lastIndex)
+		if hi > tc.cache[i].index && hi < tc.cache[i+1].index {
+			tc.cache[i].index = hi
+			tc.cache = tc.cache[i:]
+			return
+		}
+	}
+}
+
+// ScanAppend appends a list of raft entries to the TermCache
+// It is the caller's responsibility to ensure entries is a valid raftLog.
+func (tc *TermCache) ScanAppend(ctx context.Context, entries []pb.Entry) {
```

Comment:
There is probably a way to squash this function too, like the two above. Can you try?